modules/machine/mysql-ha/main.tf

```diff
@@ -0,0 +1,92 @@
+resource "juju_application" "backups_s3_integrator" {
+  name  = "backups-s3-integrator"
+  model = var.juju_model_name
+  trust = true
+
+  charm {
+    name     = "s3-integrator"
+    channel  = var.s3_integrator_charm_channel
+    base     = var.s3_integrator_charm_base
+    revision = var.s3_integrator_charm_revision
+  }
+
+  config = {
+    endpoint     = var.mysql_backup_endpoint
+    bucket       = var.mysql_backup_bucket_name
+    path         = var.juju_model_name
+    region       = var.mysql_backup_region
+    s3-uri-style = "path"
+  }
+
+  units = 1
+
+  provisioner "local-exec" {
+    # There's currently no way to wait for the charm to be idle, hence the wait-for
+    # https://github.com/juju/terraform-provider-juju/issues/202
+    command = "juju wait-for application ${self.name} --query='name==\"${self.name}\" && status==\"blocked\"'; $([ $(juju version | cut -d. -f1) = '3' ] && echo 'juju run' || echo 'juju run-action') ${self.name}/leader sync-s3-credentials access-key=${var.mysql_backup_access_key} secret-key=${var.mysql_backup_secret_key}"
+  }
+}
+
+resource "juju_application" "mysql" {
+  name  = "mysql"
+  model = var.juju_model_name
+  trust = true
+
+  charm {
+    name     = "mysql"
+    channel  = var.mysql_charm_channel
+    revision = var.mysql_charm_revision
+    base     = var.mysql_charm_base
+  }
+
+  units       = var.mysql_charm_units
+  constraints = var.mysql_charm_constraints
```

Comment:
Juju Terraform Provider provides a storage schema for applications now. It might be useful to set it as a variable.

https://registry.terraform.io/providers/juju/juju/latest/docs/resources/application#nestedatt--storage

Reply:
@amandahla , I might misunderstood it, but isn't just read-only attributes? It seems I cannot set it

Reply:
I meant that you can deploy an Application via terraform setting the storage size you need. It might be useful if someone wants a bigger storage while deploying MySQL.
Here is an example:
https://github.com/juju/terraform-provider-juju/blob/main/examples/resources/juju_application/resource.tf

Reply:
Gotcha, good tip

[a79bebd](https://github.com/canonical/terraform-modules/pull/19/commits/a79bebdb02d3a0d6c8bbd7374cfc9f4c26893886)